sor/src/main/java/com/bazaarvoice/emodb/sor/db/cql/CachingRowGroupIterator.java

```diff
@@ -0,0 +1,149 @@
+
+        final List<SoftReference<List<Row>>> softCache = Lists.newArrayList();
+        List<Row> softCacheGroup = Lists.newArrayListWithCapacity(_softCacheGroupSize);
+
+        while (rowGroup.hasNext()) {
+            if (softCacheGroup.size() == _softCacheGroupSize) {
+                softCache.add(softlyReferenced(softCacheGroup));
+                softCacheGroup = Lists.newArrayListWithCapacity(_softCacheGroupSize);
+            }
+            softCacheGroup.add(rowGroup.next());
+        }
+
+        softCache.add(softlyReferenced(softCacheGroup));
+
+        return new Iterable<Row>() {
+            @Override
+            public Iterator<Row> iterator() {
+
+                Iterator<Row> secondaryIterator =  new AbstractIterator<Row>() {
+                    private Iterator<SoftReference<List<Row>>> _softGroupsIterator = softCache.iterator();
+                    private Iterator<Row> _currentGroupIterator = null;
+                    private Iterator<Row> _sourceIterator = null;
+
+                    private Row _lastCacheRow = cache.get(cache.size()-1);
+
+                    @Override
+                    protected Row computeNext() {
+                        if (_sourceIterator != null) {
+                            // A previous softly-cached row had been lost and the remaining rows are now being served
+                            // from the source.
+                            return getNextFromSourceIterator();
+                        }
+
+                        // Check if we are currently traversing a soft cache group
+                        Row row = getNextRowFromCurrentGroupIterator();
+                        if (row != null) {
+                            return row;
+                        }
+
+                        // Try to load the next group from the soft cache
+                        if (!_softGroupsIterator.hasNext()) {
+                            return endOfData();
+                        }
+
+                        SoftReference<List<Row>> softGroupRef = _softGroupsIterator.next();
+                        List<Row> softGroup = softGroupRef.get();
+
+                        if (softGroup != null) {
+                            // Soft group has not been garbage-collected; start iterating over it.  Note this will
+                            // make the group contain a hard reference until it is fully iterated.
+                            _currentGroupIterator = softGroup.iterator();
+                            // The soft group always contains at least one record, so this will never return null
+                            return getNextRowFromCurrentGroupIterator();
+                        }
+
+                        // The soft row was garbage collected.  Only option now is to reload the remainder from
+                        // the backend.
+
+                        _sourceIterator = rowGroup.reloadRowsAfter(_lastCacheRow);
```

Comment:
Yeah, I thought about that but figured if there were memory pressures they'd be GC'd anyway.  It's a good idea; I'll make the change.